Implement Host1x nvdrv and HW emulation

bylaws opened the pull request:
See the individual commits for more info. All games using the nvhost-{nvdec, vic} interfaces should now work however they're effectively stubbed.

The actual engine implementations will be in a future PR.

PixelyIon committed:
Fix KTransferMemory Destruction Behavior + Add Missing SVC Handle Logs

KTransferMemory needs to be reset to RW on destruction unlike KSharedMemory which is simply freed on destruction, not emulating this behavior accurately leads to `deko_examples` from `switch-examples` to lead to a SEGFAULT on selecting an example as it expects the memory to be R/W while it ends up being freed instead

PixelyIon committed:
Fix `GraphicBufferProducer` Format Bug

The format provided in `GraphicBuffer` can be misleading and is supplied as `None` by the Deko3D Swapchain, it instead supplies the real format in the `NvGraphicHandle` which we now utilize instead of the one in `GraphicBuffer`.

PixelyIon committed:
Introduce `GraphicContext` and Implement Viewport Transform + Scissors

PixelyIon committed:
Introduce Boost as a submodule + Minor Fixes

Utilize Boost Container's `small_vector` for optimizing allocations, fix certain implicit casting issues and make `ILogger` not output an additional newline in the log when the application supplies one at the end of the log

PixelyIon committed:
Initial Texture Manager Implementation + Maxwell3D Render Target

Implement the groundwork for the texture manager to be able to report basic overlaps and be extended to support more in the future. The Maxwell3D registers `RenderTargetControl`, `RenderTarget` and a stub for `ClearBuffers` were implemented. 

A lot of changes were also made to `GuestTexture`/`Texture` for supporting mipmapping and multiple array layers alongside significant architectural changes to `GuestTexture` effectively disconnecting it from `Texture` with it no longer being a parent rather an object that can be used to create a `Texture` object.

Note: Support for fragmented CPU mappings hasn't been added for texture synchronization yet

PixelyIon committed:
Texture GMMU Address Resolution + Refactor `Maxwell3D::CallMethod`

Fixes bugs with the Texture Manager lookup, fixes `RenderTarget` address extraction (`low`/`high` were flipped prior), refactors `Maxwell3D::CallMethod` to utilize a case for the register being modified + preventing redundant method calls when no new value is being written to the register, and fixes the behavior of shadow RAM which was broken previously and would lead to incorrect arguments being utilized for methods.

PixelyIon committed:
Validation Layer Filter + Fix `Texture`, GPU & `PresentationEngine` bugs

This commit implements a filter by type for any validation layer output, this allows filtering out any logs which may be unnecessary and additionally triggering a breakpoint as required. 

An issue concerning the `NDEBUG` flag never being set was fixed, it's now supplied as a release compiler flag. The issue can manifest itself by always relying on a validation layer even though it shouldn't on release, this is why the validation layer was mistakenly disabled entirely previously by using `#ifndef` rather than `#ifdef`.

An issue with the initial layout of a texture being supplied as neither `VK_IMAGE_LAYOUT_UNDEFINED` or `VK_IMAGE_LAYOUT_PREINITIALIZED` was fixed, these cases are now handled by transitioning to those layouts after creating the image rather than supplying it within `initialLayout`.

Another issue was fixed regarding not maintaining a transformation after a surface has been destroyed and recreated existed and manifested itself when the user would go out of the app and come back in, they would see the surface having an identity transformation rather than the desired one.

PixelyIon committed:
Fix Command Buffer Allocation & `FenceCycle`

This commit fixes a major issue with command buffer allocation which would result in only being able to utilize a command buffer slot on the 2nd attempt to use it after it's freed, this would lead to a significantly larger amount of command buffers being created than necessary. It also fixes an issue with the command buffers not being reset after they were utilized which results in UB eventually.

Another issue was fixed with `FenceCycle` where all dependencies are only destroyed on destruction of the `FenceCycle` itself rather than the function where the `VkFence` was found to be signalled.

PixelyIon committed:
Introduce `CommandExecutor` + Implement `ClearBuffers` + More RT Formats

This commit introduces the `CommandExecutor` which is responsible for creating and orchestrating a Vulkan command graph comprising of `CommandNode`s that construct all the objects required for rendering. As a result of the infrastructure provided by `CommandExecutor`, `ClearBuffers` could be implemented and be appropriately utilized.

A bug regarding scissors was also determined and fixed in the PR, the extent of them were previously inaccurate and this has now been fixed.

Note: We don't synchronize any textures from the guest for now as this would override the contents on the host, this'll be fixed with the appropriate write tracking but it also results in a black screen for anything that writes to FB

PixelyIon committed:
Subpass Support + More RT Formats + Fix `FenceCycle` Cyclic Dependencies

Support for subpasses was added by reworking attachment reuse code to account for preserved attachments and subpass dependencies. A lot of RT formats were also added to allow SMO to boot up entirely, it should be noted that it doesn't render anything. 

`FenceCycle` had a cyclic dependency which broke clean exit, we now utilize `std::weak_ptr<FenceCycle>` inside the `Texture` object. A minor fix for broken stack traces was also made caused by supplying a `nullptr` C-string to libfmt when a symbol was unresolved which caused an `abort` due to invocation of `strlen` with it.

PixelyIon committed:
Improve Shared Fonts + Fix AM `PopLaunchParameter` & Choreographer Bug

* Move Shared Font TTFs to AAsset storage + Support external shared font loading from `/data/data/skyline.emu/data/fonts`
* Fix bug in `IApplicationFunctions::PopLaunchParameter` caused by ignoring `LaunchParameterKind`
* Fix bug with Choreographer causing it to be awoken and exit prior to the destruction of `PresentationEngine`
* Fix bug with `IDirectory::Read` where it used `inputBuf` for the output buffer rather than `outputBuf`
* Improve `GetFunctionStackTrace` logs when `dli_sname` or `dli_fname` are missing
* Support more RT Formats

PixelyIon committed:
Introduce Texture Always Sync + Wait on GPU Execution + More RT Formats

Infrastructure for always syncing textures has been introduced now, they will be synced prior to and after every execution. This does considerably reduce the performance alongside waiting on GPU execution to finish but it will be partially recouped once conditional syncing is performed.

bylaws committed:
Implement multichannel support for GPU

Allows the execution of multiple channels at the same time, with locking
being performed on the host GPU scheduler layer, address spaces can be
bound to one or more channels.

PixelyIon committed:
Introduce `ThreadLocal` Class + Fix Several GPU Bugs

* Fix `AddClearColorSubpass` bug where it would not generate a `VkCmdNextSubpass` when an attachment clear was utilized
* Fix `AddSubpass` bug where the Depth Stencil texture would not be synced
* Respect `VkCommandPool` external synchronization requirements by making it thread-local with a custom RAII wrapper
* Fix linear RT width calculation as it's provided in terms of bytes rather than format units
* Fix `AllocateStagingBuffer` bug where it would not supply `eTransferDst` as a usage flag
* Fix `AllocateMappedImage` where `VkMemoryPropertyFlags` were not respected resulting in non-`eHostVisible` memory being utilized
* Change feature requirement in `AndroidManifest.xml` to Vulkan 1.1 from OGL 3.1 as this was incorrect

bylaws committed:
Implement pre-wait and post-increment syncpoint operations in submit

These are used by both OpenGL and Vulkan games as opposed to including
the operations inside the main commandbuffer.

PixelyIon committed:
Improve Clean Exit + Audio Pausing + Improve System Language Setting

* Clean Exiting was improved by implementing a robust system for when to abandon clean exiting and simply restart the process alongside moving clean exiting to the background when the application is quit by using the back button
* Audio is now automatically paused whenever the application is moved to the background and automatically resumed when it's brought to the foreground
* The system language setting had several errors and inconsistencies which have now been fixed, it's been brought more in line with HOS language (Albeit not entirely due to no region setting in Skyline)
* Fix a bug with `ThreadLocal` where the atomic `list` pointer was uninitialized resulting in a `SEGFAULT` during the destructor

bylaws committed:
Implement support for pushbuffer methods split across multiple GpEntries

These are used heavily in OpenGL games, which now, together with the
previous syncpoint changes, work perfectly. The actual implementation is
rather novel as rather than using a per-class state machine for all
methods we only use it for those that are known to be split across
GpEntry boundaries, as a result only a single bounds check is added to
the hot path of contiguous method execution and the performance loss is
negligible.

PixelyIon committed:
Revamp `LicenseDialog` + Update Licenses + Stop Bintray Usage

* Update licenses for dependent projects
* Add copyright notices (as provided)
* Revamp styling for `LicenseDialog` 
* Fix invisible `PreferenceDialog` buttons in Settings
* Consolidating color variables into `colorPrimary`, `backgroundColor` and `backgroundColorVariant`
* Use `com.google.android.flexbox:flexbox:3.0.0` (Google Maven) rather than `com.google.android:flexbox:2.0.1` (Bintray)

PixelyIon committed:
Split monolithic `common.h` into smaller chunks

* Resolves dependency cycles in some components
* Allows for easier navigation of certain components like `span` which were especially large
* Some imports have been moved from `common.h` into their own files due to their infrequency

PixelyIon committed:
Switch to using exceptions for guest exiting

Guest-driven exiting could cause objects left on the heap due to a `std::longjmp` from high up in the host call stack, this has been fixed by introducing `ExitException` which implicitly unrolls the stack with the exception handling mechanism.

PixelyIon committed:
Update `nvdrv` perms to 11.0.0+ & Implement `nvdrv:a` service

`nvdrv:a` (For Applets) is used by some older homebrew such as SpaceNX which don't fall back to `nvdrv` (For Applications).

PixelyIon committed:
Null-Check Optional NACP before extracting application title

Not doing this can lead to the NACP being filled with invalid data and led to crashes on homebrew titles like SpaceNX.

PixelyIon committed:
Fix Git slowness incurred by using Boost as a submodule

Git became significantly slower for nearly all actions due to the inclusion of Boost as a submodule which in turn had an extremely large amount of submodules itself for every single Boost component, this especially effected clients like GitKraken which had multi-second delays in operations and became unusable due to it. The reason for this ended up being checking for modifications in the Boost submodule which has been disabled by using [`submodule.<name>.ignore`](https://git-scm.com/docs/git-config#Documentation/git-config.txt-submoduleltnamegtignore) which disables any checks on the Boost submodule for changes and fixes the prior performance degradation.

PixelyIon committed:
Address CR Comments

PixelyIon committed:
Cleanup & Use C++ Concepts in `utils.h`

bylaws committed:
Avoid crash when passing unallocated syncpoint IDs to EventWait

bylaws committed:
Introduce variable sized span support to nvdrv deserialisation

The element containing the size first needs to be saved to a save slot with Save<T, slotId>, it can then be read back later as the size of a span with SlotSizeSpan<T, slotId>. This is needed to support the Host1XChannel Submit IOCTL.

bylaws committed:
Fixup GpuChannel::SetNvmapFd to take an FD rather than an nvmap handle

PixelyIon committed:
Resume rather than relaunch when same shortcut is used

PixelyIon committed:
Resume `EmulationActivity` when launcher icon is used

Fix a bug where attempting to launch Skyline from the launcher while emulation was in progress would result in `MainActivity` launching rather than `EmulationActivity`. We always want `EmulationActivity` to stay on top of the stack and be launched whenever Skyline is resumed, no other activity should be able to run in parallel to `EmulationActivity` in any user-accessible manner.

PixelyIon committed:
Consolidate `AddAttachment` Loops + Rename `Renderpass` -> `RenderPass`

PixelyIon committed:
Update NDK to 23.1.7779620

bylaws committed:
Support passing additional constructor arguments to nvdrv devices

Needed for host1x channels which use the same class for multiple types.

bylaws committed:
Reserve host1x channel syncpoints

bylaws committed:
Stub host1x channel devices and IOCTLs

host1x channels are generally similar to GPU channels however there is only one channel for each specific class (like a GPU engine) and an address space is shared between them all.

This PR implements the simple IOCTLs with the larger ones that will depend on changes outside of nvdrv being left for future commits. This is enough to partly run oss-nvjpeg.

bylaws committed:
Use semantic wrapping for nvdrv where appropriate

bylaws committed:
Enable Wconversion and fix warnings produced

bylaws committed:
Move FlatAllocator allocation error handling to the caller

This is a prerequisite for nvmap SMMU memory management, which only frees the memory handles refer to if an allocation fails.

bylaws committed:
Hookup FlatMemoryManager for SMMU into SoC

The SMMU is used to control the mappings of peripherals such as the VIC
and NVDEC.

bylaws committed:
Implement nvmap handle pinning/unpinning

nvmap allows mapping handles into the SMMU address space through 'pins'. These are refcounted mappings that are lazily freed when required due to AS space running out. Nvidia implements this by using several MRU lists of handles in order to choose which ones to free and which ones to keep, however as we are unlikely to even run into the case of not having enough address space a naive queue approach works fine. This pin infrastructure is used by nvdrv's host1x channel implementation in order to handle mapping of both command and data buffers for submit.

PixelyIon committed:
Android 12 Support + Update Libraries + Include Khronos Validation Layer

* Fix handling `SA_EXPOSE_TAGBITS` bit being set in Android 12 `sigaction`
* Fix CMake bug using `CMAKE_INTERPROCEDURAL_OPTIMIZATION_RELEASE` when not supported causing `-fuse-ld=gold` to be emitted as a linker flag
* Support using `VIBRATOR_MANAGER_SERVICE` rather than `VIBRATOR_SERVICE` on Android 12
* Optimize Imports for Kotlin code
* Move away from deprecated APIs in Kotlin or explicitly mark where it's not possible
* Update SDK, NDK and libraries
* Enable Gradle Configuration Cache

PixelyIon committed:
Silence errors from libraries

Library headers would produce errors that are out of our control and as a result of that, we just want to ignore this. This is possible by including the offending headers as system headers, compilers don't emit any warnings arising from them. This was extended to all libraries rather than just those which currently emitted warnings for consistency's sake.

PixelyIon committed:
Fix `GPU::DebugCallback` Type String Extraction

The second parameter of `std::string_view::substr` was assumed to be an end position (similar to `std::span`) rather than `count` which it is. As a result of this, it was entirely broken but only held together by a constant factor being subtracted from it which was derived by trial and error. It's now been fixed by returning a count rather than the absolute position.

PixelyIon committed:
Fix Incorrect Scissor Extent

The decomposition from `texture::Dimension` to `vk::Rect2D` was somehow implicit and completely incorrect resulting in wrong conversion with undefined values. It's now been fixed by explicitly setting `vk::Rect2D::extent` to `scissor` specifically.

PixelyIon committed:
Fix Non-`builtin` Uncached `Vibrator` Getter

If a non-builtin vibrator was attempted to be fetched, it'd insert it in the vibrator cache and return directly as opposed to playing the vibration on it prior to returning. This has now been fixed, the value is both put into the cache and the vibration is played on it.

PixelyIon committed:
Fix OOB Vibration Array Access in `VibrateDevice`

Access to the `vibrations` field in `vibrations[3].period` could lead to UB, this has been replaced with a proper check which adds up the period over all vibrations instead. A minor cleanup with variable names and explicit types for integer arithmetic has also been done.

PixelyIon committed:
Fix Logger Message OOB Access

Certain titles can submit logs where the last field is one off by the buffer end, the logger loop now considers this and terminates if there isn't enough data left to read the field type and length.

PixelyIon committed:
Zero-Fill `IAudioRenderer::RequestUpdate` Output Buffer

Some titles don't clear the output buffer prior to submission, as the service is expected to fill all of it in, our audren implementation is incomplete and doesn't end up doing this leaving the contents of the buffer to be undefined leading to UB in the form of SEGFAULTs or the application throwing a fatal error. This has been patched over by 0-filling the buffer which is a sane default value for the fields that aren't filled in albeit not a replacement for a proper audren implementation.

PixelyIon committed:
Implement `R16G16B16A16Snorm`/`R16G16B16A16Sint` RT Formats

Utilized in "The Touryst"

PixelyIon committed:
Zero-Initialize `GuestTexture` Members

Members of `GuestTexture` were apparently not being initialized and this led to UB since they would be read as random values. Titles such as Super Mario Odyssey avoided setting `baseArrayLayer` which led to it being left at the default value which was completely random and this would lead to crashes. This commit fixes this by initializing said values correctly.

PixelyIon committed:
Implement `R16G16{Unorm/Sint/Uint}` RT Formats

Utilized in "The Touryst"

PixelyIon committed:
Use `finishAffinity` instead of `finishAndRemoveTask`

The method used to finish (`finishAndRemoveTask`) an activity prior to going back to `MainActivity` or restarting the process led to the process prematurely exiting entirely and would result in it not being restarted or another activity not being launched. This has now been fixed by utilizing `finishAffinity` in its place which correctly only ends the activities with the same affinity as the caller.

bylaws committed:
Adds bounds checks to resampler to avoid OOB reads

Resampling would sometimes perform an OOB read into `inputBuffer` due it not containing enough data to calculate corresponding the output sample, this has been fixed by introducing bounds checking to ensure that the buffer has enough data.

bylaws committed:
Fix bitmask check for event IDs > 32 in Ctrl::SyncpointFreeEventBatch

Doing 1 << 32 would result in an integer overflow rather than the desired behaviour of checking a bit, make 1 64 bit to present that.

PixelyIon committed:
Return immediately for non-joining `KProcess::Kill`s when already killed

Locking `KProcess::threadMutex` when a process is being killed by another thread with `join` can lead to the non-joining killer effectively joining as it's waiting on the joining killer to relinquish the mutex. This has been fixed by having an atomic boolean tracking if the process has already been killed and if it has, immediately returning prior to locking the mutex for any non-joining killers.

bylaws committed:
Add template utils for constructing all elements in an std::array

This avoids the excessive repetition needed for the case where array
members have no default constructor.

eg:
```c++
std::array<Type, 10> channels{util::MakeFilledArray<Type, 10>(typeConstructorArg, <...>)};
```

bylaws committed:
Avoid GetPointer on nvmap handles where they would be accessed via SMMU

GetPointer sets the sharedMemMapped flag, which should only be set if
other userspace processes have the handle mapped.

bylaws committed:
Cleanup GPFIFO comments and make Run() private

bylaws committed:
Implement the Host1x command FIFO together with barebones Host1x classes

The Host1x block of the TX1 supports 14 separate channels to which commands can be issued, these all run asynchronously so are emulated the same way as GPU channels with one FIFO emulation thread each. The command FIFO itself is very similar to the GPFIFO found in the GPU however there are some differences, mainly the introduction of classes (similar to engines) and the Mask opcode (which allows writing to a specific set of offsets much more efficiently).

There is an internal Host1x class which functions similar to the GPFIFO class in the GPU, handling general operations such as syncpoint waits, this is accessed via the simple method interface. Other channels such as NVDEC and VIC are behind the 'Tegra Host Interface' (THI) in HW, this abstracts out the classes internal details and provides a uniform method interface ontop of the Host1x method one. We emulate the THI as a templated wrapper for the underlying class.

Syncpoint increments in Host1x are different to GPU, the THI allows submitting increment requests that will be queued up and only be applied after a specific condition in the associated engine is met; however the option to for immediate increments is also available.

bylaws committed:
Fully implement the nvdrv Host1xChannel::Submit operation

This pushes a set of command buffers into the Host1x command FIFO allocated for the channel, returning fence thresholds that can be waited on for completion,

bylaws committed:
s/Host1X/Host1x/ as per Nvidia naming

bylaws committed:
Fixup nvdrv channel types to follow naming conventions

bylaws committed:
Remove unused gm20b EngineID enum

bylaws committed:
Use semantic wrapping where appropriate for class initialiser lists

bylaws committed:
Add rules on semantic wrapping and bracketed initalisation to contrib

bylaws committed:
Add virtual dtors to time service code

bylaws committed:
Remove unused virtEnd variable in FlatMemoryManager::{Read, Write}

PixelyIon committed:
Use libcxx from LLVM Project submodule

The version of libcxx shipped with Android NDK is fairly outdated and doesn't contain several features we desire such as C++ 20 ranges. This has been fixed by using libcxx directly from the LLVM Project which has been added as a submodule and can be updated independently of NDK.

PixelyIon committed:
Build libraries with `-Ofast` for debug builds

To offset some of the performance overhead of using debug builds, we now optimize all libraries using `-Ofast` while building Skyline itself with `-O0`.

PixelyIon committed:
Upgrade AGP to `7.1.0-beta01` & NDK to `24.0.7856742`

We've moved to using a beta AGP as `7.0.2` is breaks `clangd` and other C++ features on Beta/Canary Android Studio. NDK was additionally updated with `mbedtls` to fix warnings caused by it alongside some other minor fixes to code for newer versions of libcxx.

The new AGP has a bug where it does not look for executables specified in `android_gradle_build.json` in `PATH` that includes `ninja` which is provided by the `ninja-build` package on the system rather than Android SDK's CMake on GitHub Actions (Ubuntu 20.04). This has been fixed by symlinking `/usr/bin/ninja` to the project root which is searched in for the `ninja` executable.

PixelyIon committed:
Rework Joy-Con Vibration Conversion

The guest -> host vibration conversion code was entirely broken as it didn't set the vibration `start`/`end` timestamps correctly for a cycle nor did it subtract from the `totalAmplitude` (`currentAmplitude` now) when it a cycle ended due to an incorrect `if` statement and contents. It would just end up saturating the amplitude as much as possible by adding more and more to `totalAmplitude` on every cycle while never subtracting which is entirely wrong and led to a very noticeable drop in amplitude when a vibration was repeated.

It's been entirely reworked to fix all the issues listed above and remove a lot of code that had no understandable purpose. More comments have also been added to the code to make it more readable with better variable and function naming alongside it.

PixelyIon committed:
Fix `EmulationActivity.vibrateDevice` assert due to `null` Vibrator

`EmulationActivity.vibrateDevice` would assert when a `null` Vibrator is provided due to one not being set in the controller configuration. This has now been fixed by the code not playing a vibration when a vibration device isn't selected.

PixelyIon committed:
Use proper names rather than paths for submodules

Submodules were named as their relative paths prior which were not very readable, this has now been replaced with proper titles that correspond to the submodule.

PixelyIon commented on app/src/main/cpp/skyline/services/nvdrv/devices/deserialisation/deserialisation.h:
```diff
@@ -9,26 +9,37 @@
+        size_t bytes{saveSlots[ArgType::SaveSlot] * sizeof(RemoveSlotSizeSpan<ArgType>)};
+        auto out{buffer.subspan(offset, bytes).template cast<RemoveSlotSizeSpan<ArgType>, std::dynamic_extent, true>()};
+        offset += bytes;
+
+
```
One space at most

PixelyIon commented on app/src/main/cpp/skyline/services/nvdrv/devices/deserialisation/deserialisation.h:
```diff
@@ -37,23 +48,27 @@ namespace skyline::service::nvdrv::deserialisation {
         return std::tuple<Ts...>{std::forward<Ts>(ts)...};
     }
 
+
```
Same as before

PixelyIon commented on app/src/main/cpp/skyline/services/nvdrv/devices/deserialisation/deserialisation.h:
```diff
@@ -37,23 +48,27 @@ namespace skyline::service::nvdrv::deserialisation {
     template<typename Desc, typename ArgType, typename... ArgTypes>
-    constexpr auto DecodeArgumentsImpl(span<u8, Desc::Size> buffer, size_t &offset) {
+    constexpr auto DecodeArgumentsImpl(span<u8, Desc::Size> buffer, size_t &offset, std::array<size_t, NumSaveSlots> &saveSlots) {
+
```
No space here

PixelyIon commented on app/src/main/cpp/skyline/soc/host1x/classes/host1x.cpp:
```diff
@@ -0,0 +1,39 @@
+                state.logger->Debug("Increment syncpoint: {}", incrSyncpoint.index);
+                syncpoints.at(incrSyncpoint.index).Increment();
+                break;
+            }
+            case LoadSyncpointPayload32MethodId:
+                syncpointPayload = argument;
+                break;
+            case WaitSyncpoint32MethodId: {
+                u32 syncpointId{static_cast<u8>(argument)};
+                state.logger->Debug("Wait syncpoint: {}, thresh: {}", syncpointId, syncpointPayload);
+
+                syncpoints.at(syncpointId).Wait(syncpointPayload, std::chrono::steady_clock::duration::max());
+                break;
+            }
+            default:
+                state.logger->Error("Unknown host1x class method called: 0x{:X}", method);
+                break;
```
We generally have a space after each `case`, at least I do and I think it makes it more readable.

PixelyIon commented on app/src/main/cpp/skyline/soc/host1x/command_fifo.h:
```diff
@@ -0,0 +1,63 @@
+        void Start();
+
+        /**
+         * @brief Pushes a single gather into the fifo to be processed asynchronously
```
`FIFO`*

PixelyIon reviewed (changes requested):
This PR is fundamentally broken and will likely never be merged due to the major flaws denoted in the CR comments.

bylaws committed:
Remove completed nvmap TODO

Pins have been implemented so the to-do is no longer needed.

bylaws committed:
Address PR feedback

PixelyIon committed:
Upgrade AGP to `7.1.0-beta02`

Newer versions of Android Studio Beta only work correctly with `beta02` rather than `beta01` which we used prior.

PixelyIon committed:
Improve Vulkan `Texture` Synchronization

The Vulkan Pipeline Barriers were unoptimal and incorrect to some degree prior as we purely synchronized images and not staging buffers. This has now been fixed and improved in general with more relevant synchronization.

PixelyIon committed:
Introduce `JniString` as C++ wrapper over `jstring`

We used to manually call JNI UTF-8 string allocation and deallocation functions when utilizing a `jstring` but this could be erroneous and is just inconvenient. All of this has now been consolidated into an class `JniString` which is a wrapper around `std::string` and creates a copy of the contents of the `jstring` in its constructor and passes them into the `std::string` constructor.

bylaws committed:
Introduce `OffsetMember` for offset-based union members

The offset of a member in a structure was determined by its relative position and compiler alignment. This is unoptimal for larger structures such as those found in GPU Engines that are primarily named by offset rather than relative positioning and end up requiring a massive amount of padding to function as is. A solution to this problem was simply to supply member offsets directly which can be done by using `OffsetMember` alongside a `union`.

PixelyIon committed:
Extend `OffsetMember` with direct `operator=`/`operator[]`

It was found to be semantically advantageous to directly pass-through certain operators such as the assignment (`=`) and array index (`[]`) operators. These would require a dereference prior to their usage otherwise but now can be directly used.

PixelyIon committed:
Migrate `Maxwell3D::Registers` to `OffsetMember`

Maxwell3D registers were primarily written with absolute offsets and ended up being fairly messy due to attempting to emulate this using struct relative positioning resulting in a lot of pointless padding members. This has now been improved by utilizing `OffsetMember` to directly use offsets resulting in much neater code.
